content/en/docs/concepts/configuration/configmap.md

```diff
@@ -0,0 +1,78 @@
+---
+title: ConfigMaps
+content_template: templates/concept
+weight: 20
+---
+
+{{% capture overview %}}
+
+{{< glossary_definition term_id="configmap" prepend="A ConfigMap is" length="all" >}}
+
+{{< caution >}}
+ConfigMap does not provide secrecy or encryption.
+
+If the data you want to store are confidential, use a
+{{< glossary_tooltip text="Secret" term_id="secret" >}} rather than a ConfigMap,
+or use additional (third party) tools to keep your data private.
+{{< /caution >}}
+
+{{% /capture %}}
+
+{{% capture body %}}
+## ConfigMap object
```

Comment:
Expand on this section. I had the following questions:

1) What's the most common/recommended use case for a ConfigMap? (non-sensitive Environment variables?)
2) At what stage in a Pod/container's lifecycle is a ConfigMap used? (I think the answer to this is "exclusively during creation", but I'm still not sure from reading this topic)
3) Can a Pod have/reference multiple ConfigMaps? (If so what happens?)
4) Does each key in a ConfigMap's `data` element need to be unique? (presumably yes)

Reply:
I'll answer here as well as try to make the page cover these things.

>  What's the most common/recommended use case for a ConfigMap? (non-sensitive Environment variables?)

ConfigMaps are usually used to separate configuration from code. For example, you can use the same container image to run locally on your PC for testing and in the cloud for live use, and you can do that

> At what stage in a Pod/container's lifecycle is a ConfigMap used?

There are 4 ways a Pod can use a ConfigMap:
- environment variables set when, AIUI, a container for that Pod is started.
- a volume mounted into the Pod for container(s) in the Pod to read
- when the kubelet starts a container in a Pod, it can set command line arguments based on a ConfigMap
- the Pod can talk to the cluster control plane and GET / WATCH the ConfigMap directly.

I _think_ that all 4 can account for changes to the ConfigMap that happen after you start the Pod. For example, a Pod can have 2 containers, 1 that is long lives and another that exits evert few minutes. The kubelet restarts the second container and on launching the replacement containers in the Pod, the kubelet uses the _current_ ConfigMap values not the values that the kubelet saw at Pod creation time.

There is a recent alpha feature, _immutable ConfigMap_. https://blog.alcide.io/kubernetes-1.18-introduces-immutable-configmaps-and-secretes explains this quite well.

> Can a Pod have/reference multiple ConfigMaps? (If so what happens?)

Yes. For example:
```yaml
    …
      env:
        - name: FOO
          valueFrom:
            configMapKeyRef:
              name: first-configmap
              key: name-of-a-key-in-configmap-1
        - name: BAR
          valueFrom:
            configMapKeyRef:
              name: second-configmap
              key: name-of-a-key-in-configmap-2 # could also be the same key as before
```

> Does each key in a ConfigMap's data element need to be unique?

When you store an object into the Kubernetes API, you can't store duplicate keys. When you write a manifest in YAML or JSON, the effect of trying to use a duplicate key comes from the format rules of YAML or JSON.

Reply:
> the Pod can talk to the cluster control plane and GET / WATCH the ConfigMap directly.

Interesting. The question "can a ConfigMap's values change over the course of a Pod or container's lifetime, and what happens when it does?" almost made my shortlist. Might be worth answering this question, too :)